lib/services/unified_resource.go

```diff
@@ -0,0 +1,675 @@
+		return trace.BadParameter("missing parameter Key")
+	}
+	if len(replaceWith.Key) == 0 {
+		return trace.BadParameter("missing parameter Key")
+	}
+	if !bytes.Equal(expected.Key, replaceWith.Key) {
+		return trace.BadParameter("expected and replaceWith keys should match")
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	i, found := c.tree.Get(&btreeItem{Item: expected})
+	if !found {
+		return trace.CompareFailed("key %q is not found", string(expected.Key))
+	}
+	existingItem := i.Item
+	if existingItem.Value != expected.Value {
+		return trace.CompareFailed("current value does not match expected for %v", string(expected.Key))
+	}
+	event := Event{
+		Type: types.OpPut,
+		Item: replaceWith,
+	}
+	c.processEvent(event)
+	return nil
+}
+
+type getResult struct {
+	Items []Item
+}
+
+func (c *unifiedResourceCache) getRange(ctx context.Context, startKey, endKey []byte, limit int) getResult {
+	var res getResult
+	c.tree.AscendRange(&btreeItem{Item: Item{Key: startKey}}, &btreeItem{Item: Item{Key: endKey}}, func(item *btreeItem) bool {
+		res.Items = append(res.Items, item.Item)
+		if limit > 0 && len(res.Items) >= limit {
+			return false
+		}
+		return true
+	})
+	return res
+}
+
+func (c *unifiedResourceCache) processEvent(event Event) {
+	switch event.Type {
+	case types.OpPut:
+		item := &btreeItem{Item: event.Item, index: -1}
+		c.tree.ReplaceOrInsert(item)
+	case types.OpDelete:
+		item, found := c.tree.Get(&btreeItem{Item: event.Item})
+		if !found {
+			return
+		}
+		c.tree.Delete(item)
+	default:
+		// skip unsupported record
+	}
+}
+
+// UnifiedResourceWatcherConfig is a UnifiedResourceWatcher configuration.
+type UnifiedResourceWatcherConfig struct {
+	ResourceWatcherConfig
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesClusterGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// UnifiedResourceWatcher is built on top of resourceWatcher to monitor additions
+// and deletions to resources displayable in the UI such as Nodes, Dbs, Apps, etc.
+type UnifiedResourceWatcher struct {
+	*resourceWatcher
+	*unifiedResourceCollector
+}
+
+// Close closes the underlying resource watcher
+func (u *UnifiedResourceWatcher) Close() error {
+	u.resourceWatcher.Close()
+	return u.unifiedResourceCollector.Close()
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	var resources []types.ResourceWithLabels
+
+	// if the watcher is not initialized or stale, instead of returning nothing, return upstream nodes
+	if !u.IsInitialized() || u.stale {
+		nodes, err := u.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+		if err != nil {
+			return nil, trace.Wrap(err, "getting nodes while unified resource cache is uninitialized or stale")
+		}
+
+		for _, node := range nodes {
+			resources = append(resources, node)
+		}
+		return resources, nil
+	}
```

Comment:
We shouldn't omit non-node resources if it's uninitialized/stale.  Keeping consistent behavior across initialized and uninitialized state is very important (should also have test coverage to ensure that initialized and uninitialized state are indistinguishable to the caller).  Additionally, we probably don't want to be deserializing a fresh copy per incoming request.

I realize this is a non-trivial ask.  I'd suggest handling this by maintaining a separate `unifiedResourceCache` instance that isn't live (i.e. has no watcher and can be initialized immediately as-needed), and initializing it lazily as-needed, re-creating it on some TTL.  Basically, have a snapshot of a cache that you can create once and serve unlimited requests from while the primary cache is unhealthy, then reset it every 10-15s to ensure it doesn't ever get too stale.

Reply:
@zmb3 The primary cache uses a replication strategy.  In order to do that, it cannot load its initial state until after the watcher reports healthy, otherwise the cache becomes invalid and may miss put/delete events.

Falling back to a tll-based caching strategy if the primary cache is unhealthy is how `cache.Cache` works, and while its a bit heavy-weight/annoying, its proven to be fairly resilient even in the face of serious bugs/network instability.  The only other alternatives I'm aware of are:

1. Load everything for every request (same amount of work as creating a new cache per request, strictly-worse than ttl-based fallback if 2 or more requests come in).
2. Return an error if primary cache is uninitialized (flaky, poor resilience to bugs/instability).
3. Block, waiting for cache to become healthy (totally fine as long as the failure is fairly transient, but worst possible outcome if error is persistent or semi-persistent).

Note: this cache reads from `cache.Cache`, not the real backend.  Loading and building an extra cache state isn't cheap, but it's cheap enough that it's probably better than returning en error and/or blocking.